### ISSUE TITLE
Frontend: masking for more functions, global setting

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1691]
For example, currently, we have already:

* `matmul`: `use_mask`
* `reduce` and co: `use_mask`
* `moments`: `use_mask`
* `BatchNorm.__init__`: `use_mask`
* `softmax`/`log_softmax`: `use_mask`
* `pad`: `handle_dynamic_dims`
* `reverse_sequence`: `handle_dynamic_dims`

But such an option is missing for, and also not currently used for:

* `Conv1d.__init__` and co, `conv`
* `TransposedConv1d.__init__` an co, `transposed_conv`
* `pool`, `max_pool`, `pool1d` and co
* `stft`
* `window`
* `scatter`
* `sort`, ~`top_k`~ (masking is already handled), `search_sorted`
* `concat`

We have use cases for `conv`/`transposed_conv`/`pool`/`stft`, so it makes sense to implement it for them. `sort`/`top_k`, `search_sorted` is also not so trivial (maybe needs to be similar to `reduce`) but not so important for now.

However, when implementing it, what should be the default? We should not change the behavior of existing setups. So having the default `use_mask=True` would require a new behavior version. But even then, I'm not sure it is really always wanted, as there is some performance hit, and specifically for training, this is maybe not always wanted. So, despite having the default controlled via behavior version, maybe we also should introduce a new global flag? How should this be called? `rf_use_mask`? Other such global options are called eg. `rf_att_dropout_broadcast` and `rf_dropout_broadcast`.

But such `rf_use_mask`, does it make sense to have this shared for all functions? E.g. for `matmul`, `softmax`, `reduce`, `pad`, it's probably never really a good idea? For `BatchNorm.__init__`, it's maybe debatable (in the community, masking is mostly not done). `moments` unclear. For `conv` and co, this option makes sense (in the community. masking is mostly not done).

One problem is maybe, when we forget about one function now, or maybe later introduce another function where we don't properly handle masking, and then again later want to also add some `use_mask` option for this, again we need a new behavior version then to introduce it for this function. But maybe that's not really a problem.

So I would suggest now to add and implement `use_mask` for these:

* `Conv1d.__init__` and co, `conv`
* `TransposedConv1d.__init__` an co, `transposed_conv`
* `pool`, `max_pool`, `pool1d` and co
* `stft`
* `window`
* `scatter`

Basically it just means to add a `copy_masked` somewhere at the end. Maybe we can even do that in the frontend code, such that there is no special handling needed in every backend.

(For the others missing functions, it's not so straightforward. But also, at least I don't really need those at the moment. We should just make sure that there is at least some check for it, which fails with some error then.)

And then we can introduce `rf_use_mask`, which would for now control all those, and then also `BatchNorm.__init__`.

~And then also fix `merge_dims`, as discussed below ([here](https://github.com/rwth-i6/returnn/issues/1691#issuecomment-2684333551)).~ (Fixed independently via #1695.)

(cc @NeoLegends, @dorian-K)

**Comment by albertz** [comment]
While implementing this: The masking for max pooling is with -inf (as Torch max_pool does it), for avg pooling is with 0 (this is consistent to Torch avg_pool).

**Comment by albertz** [comment]
While further looking through code, I also realized that `scatter` also needs special care. Specifically, what should we do about invalid indices from the padded area? E.g. consider `rf.scatter(x, mode=..., indices=idxs, ...)`. Now consider that `idx` is [Batch,InSpatial]. The `x` should be masked accordingly.

**Comment by albertz** [comment]
Another thing: I noticed, in PyTorch `conv`, the padding we add there seems to be off. Specifically, currently we do:
```python
pad = s.dimension - 1 - (src_raw.shape[2 + i] - 1) % stride_
pad_left = pad // 2
pad_right = pad - pad_left
```

However, I noticed, I need to do this instead so that it matches when operating on a single seq:
```python
pad = s.dimension - 1
pad_left = pad // 2
pad_right = pad - pad_left - (src_raw.shape[2 + i] - 1) % stride_
```

Now I'm not exactly sure how to introduce this change. That is maybe for another new behavior version?

**Comment by albertz** [comment]
While implementing this (#1692), I stumbled upon `merge_dims`, which is a bit problematic on dynamic dims: Consider that you have dynamic `time1` and `time2`, dyn size shape [Batch], and you do `merge_dims` on them. Currently it performs dim math, resulting in `time1 * time2`, where the dyn sizes are multiplied elementwise. But that is not how the resulting tensor looks like. The dyn sizes are not correct.  We would need a dyn mask to properly describe it, but we don't have that yet. And then, when you do some op on this which applies masking, the masking would destroy the tensor. E.g. `pack_padded`/`masked_select` does this when you want to pack `time1`,`time2` into a single merged time dim: It first does `merge_dims`, and then it does `scatter`, and the `scatter` now applies masking, and thus destroys the tensor. (Luckily our test suite caught this problem.)

So, what are the solutions? We could explicitly set `use_mask=False` for this `scatter` call. That should fix the `pack_padded`/`masked_select` here.

But I'm afraid that similar code is maybe elsewhere. Basically all such usage of `merge_dims` is always problematic, specifically when we now always do proper masking. So maybe `merge_dims` should not produce such `Dim` with such incorrect dyn sizes. But this either with a flag, or this new behavior version.

**Comment by albertz** [comment]
Note, while implementing #1692, I got very bad behavior after masking with -inf for max pooling. This was fixed by masking the output with 0 again. So, whenever we mask with -inf, for certain functions (e.g. also scatter with mode=max), we should be careful.

**Comment by albertz** [comment]
Note, I realized that the config option `rf_use_mask` can maybe be extended: E.g. maybe for training, the user accepts to not use masking, but then for recognition/inference, the user would want to have masking. We could support sth like `rf_use_mask = "only_for_eval"` or so. But we don't need to implement this now. This can be easily extended later.
